llvm/lib/CodeGen/MachineBasicBlock.cpp

```diff
@@ -156,12 +156,14 @@ void ilist_traits<MachineInstr>::addNodeToList(MachineInstr *N) {
   MachineFunction *MF = Parent->getParent();
   N->addRegOperandsToUseLists(MF->getRegInfo());
   MF->handleInsertion(*N);
+  Parent->noteInsertion(*N);
```

Comment:
Is there any measurable performance impact from this? You could push to a branch tracked by https://llvm-compile-time-tracker.com/ to check.

Reply:
Also if we're going to have both per-MF and per-MBB delegates, I wonder if there is some clever way of implementing this so that we only require one `handleInsertion` call here, and the MBB handleInsertion automatically forwards to the MF handleInsertion. Or vice versa.

As you can probably tell I have not thought this through, just wondering if we can reduce the overhead when there are no MF delegates **and** no MBB delegates.

Reply:
Actually do we need a MBB delegate here at all?

Put differently why isn't `MBBSplitCriticalEdgeDelegate` a `MachineFunction` delegate?

Reply:
I have changed this to be a delegate of `MachineFunction`.

I was wary that `MachineFunction` only admits a single delegate, so we might overlap with another user.
However looking at it again, the only user I can see is GlobalISel, and this shouldn't overlap.
I guess we could always extend `MachineFunction` in the future to allow multiple delegates.